xapian-core/cluster/cluster.cc

```diff
@@ -0,0 +1,586 @@
+}
+
+DocumentSet
+Cluster::Internal::get_documents()
+{
+    DocumentSet docs;
+    for (auto&& point : cluster_docs)
+	docs.add_document(point.get_document());
+    return docs;
+}
+
+Point
+Cluster::get_index(unsigned int index) const
+{
+    LOGCALL(API, Point, "Cluster::get_index", index);
+    return internal->get_index(index);
+}
+
+Point
+Cluster::Internal::get_index(unsigned int index) const
+{
+    return cluster_docs[index];
+}
+
+ClusterSet&
+ClusterSet::operator=(const ClusterSet &other)
+{
+    // pointers are reference counted.
+    internal = other.internal;
+    return *this;
+}
+
+ClusterSet::ClusterSet(const ClusterSet &other)
+    : internal(other.internal)
+{
+}
+
+ClusterSet::ClusterSet() : internal(new Xapian::ClusterSet::Internal)
+{
+}
+
+ClusterSet::~ClusterSet()
+{
+}
+
+doccount
+ClusterSet::Internal::size() const
+{
+    return clusters.size();
+}
+
+doccount
+ClusterSet::size() const
+{
+    LOGCALL(API, doccount, "ClusterSet::size", NO_ARGS);
+    return internal->size();
+}
+
+void
+ClusterSet::Internal::add_cluster(const Cluster &c)
+{
+    clusters.push_back(c);
+}
+
+void
+ClusterSet::add_cluster(const Cluster &c)
+{
+    LOGCALL_VOID(API, "ClusterSet::add_cluster", c);
+    internal->add_cluster(c);
+}
+
+Cluster&
+ClusterSet::Internal::get_cluster(doccount i)
+{
+    return clusters[i];
+}
+
+Cluster&
+ClusterSet::operator[](doccount i)
+{
+    return internal->get_cluster(i);
+}
+
+void
+ClusterSet::Internal::add_to_cluster(const Point &x, unsigned int i)
+{
+    clusters[i].add_point(x);
+}
+
+void
+ClusterSet::add_to_cluster(const Point &x, unsigned int i)
+{
+   LOGCALL_VOID(API, "ClusterSet::add_to_cluster", x | i);
+   internal->add_to_cluster(x, i);
+}
+
+void
+ClusterSet::Internal::recalculate_centroids()
+{
+    for (vector<Cluster>::iterator it = clusters.begin(); it != clusters.end(); ++it)
+	(*it).recalculate();
+}
+
+void
+ClusterSet::recalculate_centroids()
+{
+    LOGCALL_VOID(API, "ClusterSet::recalculate_centroids", NO_ARGS);
+    internal->recalculate_centroids();
+}
+
+void
+ClusterSet::clear_clusters()
+{
+    LOGCALL_VOID(API, "ClusterSet::clear_clusters", NO_ARGS);
+    internal->clear_clusters();
+}
+
+void
+ClusterSet::Internal::clear_clusters()
+{
+    for (auto&& cluster : clusters)
+	cluster.clear();
+}
+
+doccount
+Cluster::size() const
+{
+    LOGCALL(API, doccount, "Cluster::size", NO_ARGS);
+    return internal->size();
+}
+
+doccount
+Cluster::Internal::size() const
+{
+    return (cluster_docs.size());
+}
+
+void
+Cluster::add_point(const Point &doc)
+{
+    LOGCALL_VOID(API, "Cluster::add_point", doc);
+    internal->add_point(doc);
+}
+
+void
+Cluster::Internal::add_point(const Point &doc)
+{
+    cluster_docs.push_back(doc);
+}
+
+void
+Cluster::clear()
+{
+    LOGCALL_VOID(API, "Cluster::clear", NO_ARGS);
+    internal->clear();
+}
+
+void
+Cluster::Internal::clear()
+{
+    cluster_docs.clear();
+}
+
+Centroid&
+Cluster::get_centroid() const
```

Comment:
Yes. This should probably apply to all other const methods which are returning references.